runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,346 @@
+        type = protocol.FileChangeType.Created
+      end
+    end
+
+    local matches_filter = false
+    local filters = watched_dirs[dir].callbacks[client_id][reg_id].filters
+    for _, filter in ipairs(filters) do
+      if match(filter.pattern, path) and math.floor(filter.kind / (2^(type-1))) % 2 == 1 then
+        matches_filter = true
+        break
+      end
+    end
+    if not matches_filter then return end
+
+    local change = {
+      uri = vim.uri_from_fname(path),
+      type = type,
+    }
+    change_cache[client_id] = change_cache[client_id] or {}
+    local last_type = change_cache[client_id][change.uri]
+    if not last_type or last_type ~= change.type then
+      change_queue[client_id] = change_queue[client_id] or {}
+      table.insert(change_queue[client_id], change)
+      change_cache[client_id][change.uri] = change.type
+    end
+    if not queue_timers[client_id] then
+      local t = uv.new_timer()
+      t:start(100, 0, function()
+        t:close()
+        vim.lsp.get_client_by_id(client_id).notify("workspace/didChangeWatchedFiles", {
+          changes = change_queue[client_id],
+        })
+        change_queue[client_id] = nil
+        change_cache[client_id] = nil
+        queue_timers[client_id] = nil
+      end)
+      queue_timers[client_id] = t
+    end
+  end
+end
+
+-- Never create watchers for directories matching these patterns. Similar
+-- to vscode's files.watcherExclude setting.
+local excludes = {
+  parse("**/.git/objects/**"),
+  parse("**/.git/subtree-cache/**"),
+  parse("**/node_modules/*/**"),
+  parse("**/.hg/store/**"),
+}
+
+---@private
+--- Creates libuv fs_events handles.
+---
+---@param dir string Absolute path to the directory to watch.
+---@param pattern string|table The LSP glob pattern (raw or parsed) to match against.
+---@param kind number The LSP WatchKind value.
+---@param client_id number The LSP client's ID.
+---@param reg_id string The ID used to register the request.
+local function fsevent_ensure_recursive(dir, pattern, kind, client_id, reg_id)
+  -- Recursive flag for libuv watcher not implemented on linux
+  local recursive_watch = vim.fn.has("mac") == 1 or vim.fn.has("win32") == 1
+
+  if not watched_dirs[dir] then
+    local fsevent, fserr = uv.new_fs_event()
+    assert(not fserr, fserr)
+    watched_dirs[dir] = {
+      fsevent = fsevent,
+      callbacks = {},
+    }
+    fsevent:start(dir, { recursive = recursive_watch }, function(err, filename, events)
+      assert(not err, err)
+      for _, reg in pairs(watched_dirs[dir].callbacks[client_id]) do
+        reg.callback(filename, events)
+      end
+    end)
+  end
+
+  watched_dirs[dir].callbacks[client_id] = watched_dirs[dir].callbacks[client_id] or {}
+  watched_dirs[dir].callbacks[client_id][reg_id] = watched_dirs[dir].callbacks[client_id][reg_id] or {}
+  watched_dirs[dir].callbacks[client_id][reg_id].callback = watched_dirs[dir].callbacks[client_id][reg_id].callback or get_callback(dir, client_id, reg_id)
```

Comment:
I understand this kind of pattern to initialize nested objects is very ugly, so any pointers to clean this up are most welcome!

Reply:
Maybe `vim.defaulttable()` could help?